File: monitoring/exporters/health-check.py
"""
Health Check Middleware для Python приложений (FastAPI/Flask/Django)
Предоставляет эндпоинты для мониторинга состояния сервисов
"""

import asyncio
import time
import psutil
import os
import json
import logging
from typing import Dict, Any, Optional
from datetime import datetime, timezone
from prometheus_client import Counter, Histogram, Gauge, generate_latest, CONTENT_TYPE_LATEST

try:
    from fastapi import FastAPI, Request, Response
    from fastapi.responses import JSONResponse
    FASTAPI_AVAILABLE = True
except ImportError:
    FASTAPI_AVAILABLE = False

try:
    from flask import Flask, request, jsonify, Response
    FLASK_AVAILABLE = True
except ImportError:
    FLASK_AVAILABLE = False

try:
    import asyncpg
    import aioredis
    import httpx
    ASYNC_DEPS_AVAILABLE = True
except ImportError:
    ASYNC_DEPS_AVAILABLE = False

# Prometheus метрики
http_requests_total = Counter('http_requests_total', 'Total number of HTTP requests', 
                             ['method', 'endpoint', 'status_code'])

http_request_duration = Histogram('http_request_duration_seconds', 'HTTP request duration',
                                 ['method', 'endpoint'])

app_uptime = Gauge('app_uptime_seconds', 'Application uptime in seconds')

system_memory_usage = Gauge('system_memory_usage_bytes', 'System memory usage in bytes', 
                           ['type'])

system_cpu_usage = Gauge('system_cpu_usage_percent', 'System CPU usage percentage')

class HealthCheckMiddleware:
    """Основной класс для проверки состояния системы"""
    
    def __init__(self, config: Dict[str, Any] = None):
        self.config = {
            'health_check_path': '/health',
            'ready_check_path': '/ready',
            'live_check_path': '/live',
            'metrics_path': '/metrics',
            'detailed_checks': True,
            'timeout': 5,
            'external_apis': {},
            'database_url': None,
            'redis_url': None,
            'log_level': 'INFO',
            **config or {}
        }
        
        self.start_time = time.time()
        self.logger = self._setup_logger()
        
        # Асинхронные клиенты (будут инициализированы позже)
        self.db_pool: Optional[object] = None
        self.redis_client: Optional[object] = None
        self.http_client: Optional[object] = None
        
        # Обновление системных метрик
        asyncio.create_task(self._update_system_metrics())
        
    def _setup_logger(self) -> logging.Logger:
        """Настройка логгера"""
        logger = logging.getLogger('health_check')
        logger.setLevel(getattr(logging, self.config['log_level']))
        
        if not logger.handlers:
            handler = logging.StreamHandler()
            formatter = logging.Formatter(
                '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
            )
            handler.setFormatter(formatter)
            logger.addHandler(handler)
            
        return logger
    
    async def initialize_async_clients(self):
        """Инициализация асинхронных клиентов"""
        if not ASYNC_DEPS_AVAILABLE:
            return
            
        # Инициализация пула соединений с БД
        if self.config['database_url']:
            try:
                self.db_pool = await asyncpg.create_pool(
                    self.config['database_url'],
                    min_size=1,
                    max_size=5,
                    command_timeout=60
                )
            except Exception as e:
                self.logger.error(f"Failed to initialize database pool: {e}")
        
        # Инициализация Redis клиента
        if self.config['redis_url']:
            try:
                self.redis_client = await aioredis.from_url(
                    self.config['redis_url'],
                    decode_responses=True
                )
            except Exception as e:
                self.logger.error(f"Failed to initialize Redis client: {e}")
        
        # Инициализация HTTP клиента
        self.http_client = httpx.AsyncClient(timeout=self.config['timeout'])
    
    async def health_check(self, request: Request = None) -> Dict[str, Any]:
        """Основной health check endpoint"""
        try:
            checks = await self.perform_health_checks()
            status = self.determine_overall_status(checks)
            
            response = {
                'status': status,
                'timestamp': datetime.now(timezone.utc).isoformat(),
                'version': os.getenv('APP_VERSION', '1.0.0'),
                'uptime': int(time.time() - self.start_time),
                'checks': self.config['detailed_checks'] else checks
            }
            
            return JSONResponse(
                content=response,
                status_code=200 if status == 'healthy' else 503
            )
            
        except Exception as e:
            self.logger.error(f"Health check failed: {e}")
            return JSONResponse(
                content={
                    'status': 'unhealthy',
                    'error': str(e),
                    'timestamp': datetime.now(timezone.utc).isoformat()
                },
                status_code=503
            )
    
    async def ready_check(self, request: Request = None) -> Dict[str, Any]:
        """Ready check endpoint (Kubernetes)"""
        ready = await self.check_readiness()
        
        response = {
            'ready': ready,
            'timestamp': datetime.now(timezone.utc).isoformat(),
            'checks': {
                'database': await self.check_database_connection(),
                'redis': await self.check_redis_connection(),
                'external_apis': await self.check_external_apis()
            }
        }
        
        return JSONResponse(
            content=response,
            status_code=200 if ready else 503
        )
    
    async def live_check(self, request: Request = None) -> Dict[str, Any]:
        """Live check endpoint (Kubernetes)"""
        response = {
            'live': True,
            'timestamp': datetime.now(timezone.utc).isoformat(),
            'uptime': int(time.time() - self.start_time)
        }
        
        return JSONResponse(content=response, status_code=200)
    
    async def metrics_endpoint(self, request: Request = None) -> Response:
        """Prometheus metrics endpoint"""
        return Response(
            generate_latest(),
            media_type=CONTENT_TYPE_LATEST
        )
    
    async def perform_health_checks(self) -> Dict[str, Any]:
        """Выполнение всех проверок состояния"""
        checks = {}
        
        # Системные проверки
        checks['system'] = await self.check_system()
        
        # База данных
        checks['database'] = await self.check_database_connection()
        
        # Redis
        checks['redis'] = await self.check_redis_connection()
        
        # Внешние API
        checks['external_apis'] = await self.check_external_apis()
        
        # Память
        checks['memory'] = await self.check_memory_usage()
        
        # Дисковое пространство
        checks['disk_space'] = await self.check_disk_space()
        
        return checks
    
    async def check_readiness(self) -> bool:
        """Проверка готовности сервиса"""
        db_ok = await self.check_database_connection()
        redis_ok = await self.check_redis_connection()
        
        return db_ok['status'] and redis_ok['status']
    
    async def check_system(self) -> Dict[str, Any]:
        """Проверка системных ресурсов"""
        cpu_percent = psutil.cpu_percent(interval=1)
        memory = psutil.virtual_memory()
        load_avg = os.getloadavg() if hasattr(os, 'getloadavg') else [0, 0, 0]
        
        healthy = (
            cpu_percent < 80 and
            memory.percent < 85 and
            load_avg[0] < 2.0
        )
        
        return {
            'status': healthy,
            'data': {
                'cpu_usage_percent': cpu_percent,
                'memory_usage_percent': memory.percent,
                'load_average': list(load_avg),
                'free_memory_gb': round(memory.free / (1024**3), 2),
                'total_memory_gb': round(memory.total / (1024**3), 2)
            }
        }
    
    async def check_database_connection(self) -> Dict[str, Any]:
        """Проверка подключения к базе данных"""
        if not self.db_pool:
            return {'status': False, 'error': 'Database not configured'}
        
        try:
            start = time.time()
            async with self.db_pool.acquire() as conn:
                await conn.fetchval('SELECT 1')
            duration = (time.time() - start) * 1000
            
            healthy = duration < 1000  # Response time should be < 1 second
            
            return {
                'status': healthy,
                'duration_ms': round(duration, 2),
                'pool_info': await self.get_pool_info()
            }
        except Exception as e:
            return {
                'status': False,
                'error': str(e)
            }
    
    async def check_redis_connection(self) -> Dict[str, Any]:
        """Проверка подключения к Redis"""
        if not self.redis_client:
            return {'status': False, 'error': 'Redis not configured'}
        
        try:
            start = time.time()
            await self.redis_client.ping()
            duration = (time.time() - start) * 1000
            
            healthy = duration < 500  # Response time should be < 0.5 seconds
            
            return {
                'status': healthy,
                'duration_ms': round(duration, 2),
                'info': await self.redis_client.info()
            }
        except Exception as e:
            return {
                'status': False,
                'error': str(e)
            }
    
    async def check_external_apis(self) -> Dict[str, Any]:
        """Проверка внешних API"""
        apis = self.config['external_apis']
        results = {}
        
        for name, config in apis.items():
            try:
                start = time.time()
                response = await self.http_client.get(config['url'])
                duration = (time.time() - start) * 1000
                
                healthy = 200 <= response.status_code < 300
                
                results[name] = {
                    'status': healthy,
                    'duration_ms': round(duration, 2),
                    'status_code': response.status_code
                }
            except Exception as e:
                results[name] = {
                    'status': False,
                    'error': str(e)
                }
        
        all_healthy = not apis or all(result['status'] for result in results.values())
        
        return {
            'status': all_healthy,
            'apis': results
        }
    
    async def check_memory_usage(self) -> Dict[str, Any]:
        """Проверка использования памяти"""
        process = psutil.Process()
        memory_info = process.memory_info()
        virtual_memory = psutil.virtual_memory()
        
        heap_usage_percent = (
            memory_info.rss / virtual_memory.total
        ) * 100
        
        healthy = heap_usage_percent < 80
        
        return {
            'status': healthy,
            'data': {
                'heap_usage_percent': round(heap_usage_percent, 2),
                'heap_used_mb': round(memory_info.rss / (1024**2), 2),
                'heap_total_mb': round(memory_info.rss / (1024**2), 2),
                'virtual_memory_percent': virtual_memory.percent
            }
        }
    
    async def check_disk_space(self) -> Dict[str, Any]:
        """Проверка дискового пространства"""
        disk_usage = psutil.disk_usage('/')
        usage_percent = (disk_usage.used / disk_usage.total) * 100
        
        healthy = usage_percent < 85
        
        return {
            'status': healthy,
            'data': {
                'usage_percent': round(usage_percent, 2),
                'free_gb': round(disk_usage.free / (1024**3), 2),
                'total_gb': round(disk_usage.total / (1024**3), 2)
            }
        }
    
    async def _update_system_metrics(self):
        """Обновление системных метрик"""
        while True:
            try:
                # Обновление uptime
                app_uptime.set(time.time() - self.start_time)
                
                # Обновление памяти
                memory = psutil.virtual_memory()
                system_memory_usage.labels(type='total').set(memory.total)
                system_memory_usage.labels(type='used').set(memory.used)
                system_memory_usage.labels(type='free').set(memory.free)
                system_memory_usage.labels(type='available').set(memory.available)
                
                # Обновление CPU
                cpu_percent = psutil.cpu_percent()
                system_cpu_usage.set(cpu_percent)
                
            except Exception as e:
                self.logger.error(f"Failed to update system metrics: {e}")
            
            await asyncio.sleep(5)
    
    async def get_pool_info(self) -> Dict[str, Any]:
        """Получение информации о пуле соединений"""
        if not self.db_pool:
            return {'error': 'No database pool'}
        
        try:
            return {
                'size': self.db_pool.get_size(),
                'min_size': self.db_pool.get_min_size(),
                'max_size': self.db_pool.get_max_size(),
                'closed': self.db_pool.is_closing()
            }
        except Exception as e:
            return {'error': str(e)}
    
    def determine_overall_status(self, checks: Dict[str, Any]) -> str:
        """Определение общего статуса системы"""
        for check in checks.values():
            if isinstance(check, dict) and check.get('status') is False:
                return 'unhealthy'
        
        for check in checks.values():
            if isinstance(check, dict) and check.get('status') == 'warning':
                return 'degraded'
        
        return 'healthy'

# FastAPI интеграция
if FASTAPI_AVAILABLE:
    def create_fastapi_health_middleware(config: Dict[str, Any] = None) -> FastAPI:
        """Создание FastAPI приложения с health check endpoints"""
        
        app = FastAPI(title="Health Check Service")
        health_check = HealthCheckMiddleware(config)
        
        @app.on_event("startup")
        async def startup_event():
            await health_check.initialize_async_clients()
        
        # Health check endpoints
        app.add_api_route(
            health_check.config['health_check_path'],
            health_check.health_check,
            methods=["GET"]
        )
        
        app.add_api_route(
            health_check.config['ready_check_path'],
            health_check.ready_check,
            methods=["GET"]
        )
        
        app.add_api_route(
            health_check.config['live_check_path'],
            health_check.live_check,
            methods=["GET"]
        )
        
        app.add_api_route(
            health_check.config['metrics_path'],
            health_check.metrics_endpoint,
            methods=["GET"]
        )
        
        return app

# Flask интеграция  
if FLASK_AVAILABLE:
    def create_flask_health_middleware(config: Dict[str, Any] = None):
        """Создание Flask приложения с health check endpoints"""
        
        app = Flask(__name__)
        health_check = HealthCheckMiddleware(config)
        
        @app.before_first_request
        async def initialize():
            await health_check.initialize_async_clients()
        
        # Health check endpoints
        @app.route(health_check.config['health_check_path'], methods=['GET'])
        async def health():
            return await health_check.health_check()
        
        @app.route(health_check.config['ready_check_path'], methods=['GET'])
        async def ready():
            return await health_check.ready_check()
        
        @app.route(health_check.config['live_check_path'], methods=['GET'])
        async def live():
            return await health_check.live_check()
        
        @app.route(health_check.config['metrics_path'], methods=['GET'])
        async def metrics():
            return await health_check.metrics_endpoint()
        
        return app

# Пример использования
if __name__ == "__main__":
    import uvicorn
    
    # FastAPI пример
    if FASTAPI_AVAILABLE:
        config = {
            'database_url': 'postgresql://user:pass@localhost:5432/db',
            'redis_url': 'redis://localhost:6379',
            'external_apis': {
                'payment_api': {'url': 'http://payment-service:8080/health'},
                'notification_api': {'url': 'http://notification-service:8080/health'}
            }
        }
        
        app = create_fastapi_health_middleware(config)
        uvicorn.run(app, host="0.0.0.0", port=8000)

File: monitoring/exporters/health-check.js
/**
 * Health Check Middleware для Node.js приложений
 * Предоставляет эндпоинты для мониторинга состояния сервисов
 */

const express = require('express');
const prometheus = require('prom-client');
const { createPool } = require('mysql2/promise');
const redis = require('redis');

class HealthCheckMiddleware {
  constructor(config = {}) {
    this.config = {
      healthCheckPath: '/health',
      readyCheckPath: '/ready',
      liveCheckPath: '/live',
      metricsPath: '/metrics',
      detailedChecks: true,
      timeout: 5000,
      externalAPIs: {},
      ...config
    };

    // Prometheus метрики
    this.initializeMetrics();
    
    // Подключения
    this.database = config.database;
    this.redis = config.redis;
    
    // Метрики времени работы
    this.startTime = Date.now();
  }

  /**
   * Инициализация Prometheus метрик
   */
  initializeMetrics() {
    // Счетчик запросов
    this.requestCounter = new prometheus.Counter({
      name: 'http_requests_total',
      help: 'Total number of HTTP requests',
      labelNames: ['method', 'route', 'status_code']
    });

    // Гистограмма времени ответа
    this.responseTime = new prometheus.Histogram({
      name: 'http_request_duration_seconds',
      help: 'Duration of HTTP requests in seconds',
      labelNames: ['method', 'route'],
      buckets: [0.1, 0.3, 0.5, 0.7, 1, 3, 5, 7, 10]
    });

    // Gauge метрики
    this.uptimeGauge = new prometheus.Gauge({
      name: 'app_uptime_seconds',
      help: 'Application uptime in seconds'
    });

    this.memoryUsageGauge = new prometheus.Gauge({
      name: 'app_memory_usage_bytes',
      help: 'Application memory usage in bytes',
      labelNames: ['type']
    });

    this.cpuUsageGauge = new prometheus.Gauge({
      name: 'app_cpu_usage_percent',
      help: 'Application CPU usage percentage'
    });

    // Обновление метрик каждую секунду
    setInterval(() => {
      this.updateSystemMetrics();
    }, 1000);

    // Обновление uptime метрики
    setInterval(() => {
      this.uptimeGauge.set((Date.now() - this.startTime) / 1000);
    }, 5000);
  }

  /**
   * Express middleware для мониторинга запросов
   */
  monitoringMiddleware() {
    return (req, res, next) => {
      const start = Date.now();
      
      res.on('finish', () => {
        const duration = (Date.now() - start) / 1000;
        
        this.requestCounter
          .labels(req.method, req.route?.path || req.path, res.statusCode)
          .inc();
          
        this.responseTime
          .labels(req.method, req.route?.path || req.path)
          .observe(duration);
      });
      
      next();
    };
  }

  /**
   * Health check endpoint
   */
  async health() {
    const checks = await this.performHealthChecks();
    const status = this.determineOverallStatus(checks);
    const httpStatus = status === 'healthy' ? 200 : 503;

    const response = {
      status,
      timestamp: new Date().toISOString(),
      version: process.env.npm_package_version || '1.0.0',
      uptime: Math.floor((Date.now() - this.startTime) / 1000),
      checks: this.config.detailedChecks ? checks : undefined
    };

    return res.status(httpStatus).json(response);
  }

  /**
   * Ready check endpoint (Kubernetes)
   */
  async ready() {
    const ready = await this.checkReadiness();
    
    const response = {
      ready,
      timestamp: new Date().toISOString(),
      checks: {
        database: await this.checkDatabaseConnection(),
        redis: await this.checkRedisConnection(),
        external_apis: await this.checkExternalAPIs()
      }
    };

    return res.status(ready ? 200 : 503).json(response);
  }

  /**
   * Live check endpoint (Kubernetes)
   */
  async live() {
    const response = {
      live: true,
      timestamp: new Date().toISOString(),
      uptime: Math.floor((Date.now() - this.startTime) / 1000)
    };

    return res.status(200).json(response);
  }

  /**
   * Prometheus metrics endpoint
   */
  async metrics() {
    res.set('Content-Type', prometheus.register.contentType);
    return res.end(await prometheus.register.metrics());
  }

  /**
   * Выполнение всех проверок состояния
   */
  async performHealthChecks() {
    const checks = {};

    // Системные проверки
    checks.system = await this.checkSystem();
    
    // База данных
    checks.database = await this.checkDatabaseConnection();
    
    // Redis
    checks.redis = await this.checkRedisConnection();
    
    // Внешние API
    checks.external_apis = await this.checkExternalAPIs();
    
    // Память
    checks.memory = await this.checkMemoryUsage();
    
    // Дисковое пространство
    checks.disk_space = await this.checkDiskSpace();

    return checks;
  }

  /**
   * Проверка готовности сервиса
   */
  async checkReadiness() {
    const [dbOk, redisOk] = await Promise.all([
      this.checkDatabaseConnection(),
      this.checkRedisConnection()
    ]);
    
    return dbOk.status && redisOk.status;
  }

  /**
   * Проверка системных ресурсов
   */
  async checkSystem() {
    const os = require('os');
    const loadAvg = os.loadavg();
    const cpuUsage = await this.getCpuUsage();
    const memoryUsage = this.getMemoryUsage();
    
    const healthy = (
      loadAvg[0] < 2.0 && // CPU load average (1 min)
      cpuUsage < 80 &&    // CPU usage %
      memoryUsage < 85    // Memory usage %
    );

    return {
      status: healthy,
      data: {
        load_average: loadAvg,
        cpu_usage_percent: cpuUsage,
        memory_usage_percent: memoryUsage,
        free_memory_gb: (os.freemem() / 1024 / 1024 / 1024).toFixed(2),
        total_memory_gb: (os.totalmem() / 1024 / 1024 / 1024).toFixed(2)
      }
    };
  }

  /**
   * Проверка подключения к базе данных
   */
  async checkDatabaseConnection() {
    try {
      if (!this.database) {
        return { status: false, error: 'Database not configured' };
      }
      
      const start = Date.now();
      await this.database.query('SELECT 1');
      const duration = Date.now() - start;
      
      const healthy = duration < 1000; // Response time should be < 1 second
      
      return {
        status: healthy,
        duration: duration,
        connection_pool: await this.getConnectionPoolInfo()
      };
    } catch (error) {
      return {
        status: false,
        error: error.message
      };
    }
  }

  /**
   * Проверка подключения к Redis
   */
  async checkRedisConnection() {
    try {
      if (!this.redis) {
        return { status: false, error: 'Redis not configured' };
      }
      
      const start = Date.now();
      await this.redis.ping();
      const duration = Date.now() - start;
      
      const healthy = duration < 500; // Response time should be < 0.5 seconds
      
      return {
        status: healthy,
        duration: duration,
        info: await this.redis.info()
      };
    } catch (error) {
      return {
        status: false,
        error: error.message
      };
    }
  }

  /**
   * Проверка внешних API
   */
  async checkExternalAPIs() {
    const apis = this.config.externalAPIs;
    const results = {};
    const apiPromises = [];

    for (const [name, config] of Object.entries(apis)) {
      apiPromises.push(
        this.checkApi(name, config).then(result => ({ name, result }))
      );
    }

    const apiResults = await Promise.allSettled(apiPromises);
    
    for (const { name, result } of apiResults) {
      if (result.status === 'fulfilled') {
        results[name] = result.value;
      } else {
        results[name] = {
          status: false,
          error: result.reason.message
        };
      }
    }

    const allHealthy = Object.keys(results).length === 0 || 
      Object.values(results).every(result => result.status);

    return {
      status: allHealthy,
      apis: results
    };
  }

  /**
   * Проверка конкретного API
   */
  async checkApi(name, config) {
    const axios = require('axios');
    
    try {
      const start = Date.now();
      const response = await axios.get(config.url, {
        timeout: config.timeout || 5000,
        validateStatus: () => true
      });
      const duration = Date.now() - start;
      
      const healthy = response.status >= 200 && response.status < 300;
      
      return {
        status: healthy,
        duration: duration,
        status_code: response.status
      };
    } catch (error) {
      return {
        status: false,
        error: error.message
      };
    }
  }

  /**
   * Проверка использования памяти
   */
  async checkMemoryUsage() {
    const memUsage = process.memoryUsage();
    const totalHeap = memUsage.heapTotal;
    const usedHeap = memUsage.heapUsed;
    const external = memUsage.external;
    
    const heapPercent = (usedHeap / totalHeap) * 100;
    const totalPercent = ((usedHeap + external) / require('os').totalmem()) * 100;
    
    const healthy = heapPercent < 80;

    return {
      status: healthy,
      data: {
        heap_usage_percent: heapPercent.toFixed(2),
        total_usage_percent: totalPercent.toFixed(2),
        heap_used_mb: (usedHeap / 1024 / 1024).toFixed(2),
        heap_total_mb: (totalHeap / 1024 / 1024).toFixed(2),
        external_mb: (external / 1024 / 1024).toFixed(2)
      }
    };
  }

  /**
   * Проверка дискового пространства
   */
  async checkDiskSpace() {
    const fs = require('fs').promises;
    const os = require('os');
    
    try {
      const stats = await fs.stat(__dirname);
      const { available, total } = require('fs').statSync(__dirname);
      
      const freeBytes = available;
      const totalBytes = total;
      const usagePercent = ((totalBytes - freeBytes) / totalBytes) * 100;
      
      const healthy = usagePercent < 85;

      return {
        status: healthy,
        data: {
          usage_percent: usagePercent.toFixed(2),
          free_gb: (freeBytes / 1024 / 1024 / 1024).toFixed(2),
          total_gb: (totalBytes / 1024 / 1024 / 1024).toFixed(2)
        }
      };
    } catch (error) {
      return {
        status: false,
        error: error.message
      };
    }
  }

  /**
   * Обновление системных метрик
   */
  async updateSystemMetrics() {
    const os = require('os');
    const memUsage = process.memoryUsage();
    
    this.memoryUsageGauge.labels('heap_used').set(memUsage.heapUsed);
    this.memoryUsageGauge.labels('heap_total').set(memUsage.heapTotal);
    this.memoryUsageGauge.labels('external').set(memUsage.external);
    this.memoryUsageGauge.labels('rss').set(memUsage.rss);
    
    this.cpuUsageGauge.set(await this.getCpuUsage());
  }

  /**
   * Получение использования CPU
   */
  async getCpuUsage() {
    const startUsage = process.cpuUsage();
    await new Promise(resolve => setTimeout(resolve, 100));
    const endUsage = process.cpuUsage(startUsage);
    
    const totalUsage = endUsage.user + endUsage.system;
    const cpuUsage = (totalUsage / 100000) * 100; // Конвертация в проценты
    
    return Math.min(cpuUsage, 100);
  }

  /**
   * Получение использования памяти
   */
  getMemoryUsage() {
    const memUsage = process.memoryUsage();
    const totalMem = require('os').totalmem();
    const usagePercent = ((memUsage.heapUsed + memUsage.external) / totalMem) * 100;
    
    return Math.min(usagePercent, 100);
  }

  /**
   * Получение информации о пуле соединений
   */
  async getConnectionPoolInfo() {
    try {
      if (this.database.pool) {
        const [rows] = await this.database.pool.query('SHOW STATUS LIKE "Threads_connected"');
        return {
          active_connections: rows[0].Value,
          max_connections: 'config'
        };
      }
      return { error: 'No connection pool available' };
    } catch (error) {
      return { error: error.message };
    }
  }

  /**
   * Определение общего статуса системы
   */
  determineOverallStatus(checks) {
    // Если есть критические ошибки
    for (const check of Object.values(checks)) {
      if (check.status === false) {
        return 'unhealthy';
      }
    }
    
    // Если есть предупреждения
    for (const check of Object.values(checks)) {
      if (check.status === 'warning') {
        return 'degraded';
      }
    }
    
    return 'healthy';
  }

  /**
   * Получение Express router
   */
  getRouter() {
    const router = express.Router();
    
    // Мониторинг запросов
    router.use(this.monitoringMiddleware());
    
    // Health check endpoints
    router.get(this.config.healthCheckPath, this.health.bind(this));
    router.get(this.config.readyCheckPath, this.ready.bind(this));
    router.get(this.config.liveCheckPath, this.live.bind(this));
    router.get(this.config.metricsPath, this.metrics.bind(this));
    
    return router;
  }
}

module.exports = HealthCheckMiddleware;

// Пример использования
if (require.main === module) {
  const app = express();
  const healthCheck = new HealthCheckMiddleware({
    externalAPIs: {
      'payment_api': { url: 'http://payment-service:8080/health' },
      'notification_api': { url: 'http://notification-service:8080/health' }
    }
  });
  
  app.use(healthCheck.getRouter());
  
  app.get('/', (req, res) => {
    res.json({ message: 'Service is running' });
  });
  
  const PORT = process.env.PORT || 3000;
  app.listen(PORT, () => {
    console.log(`Health check service running on port ${PORT}`);
  });
}